### ISSUE TITLE
Render PDF to SVG

### ISSUE THREAD

**Issue by ZZYSonny** [pdf-rs/pdf#149]
Hello, recently I was writing a PDF viewer that would work exactly as a normal web page, and you can enjoy all the feature of a browser (such as awesome touchpad scrolling/zooming). Try it out here https://zzysonny.github.io/StarShooter/

The way I am rendering a page is to generate a SVG picture for each page, using mupdf. Then Chrome can directly and efficiently (when generated in a proper way) render a complex SVG file. When you zoom in using touchpad, Chrome can rerender at a higher resolution instantly.

Currently I am modifying mupdf to achieve an efficiently-renderable pdf, but that is painful and I am considering a rust pdf library. Will you be interested in implementing a PDF -> SVG renderer?

**Comment by ZZYSonny** [comment]
It seems like pdf_render is not far away from generating a svg. I will see if I can do it.

**Comment by s3bk** [comment]
It can totally generate a SVG and you can load the SVG to render it..
But why take the extra step?

**Comment by ZZYSonny** [comment]
Really? I've only seen pdf2img converting pdf to a rasterized picture.

**Comment by s3bk** [comment]
The debug print for `Scene` is SVG. Not optimized at all, but it does work.

**Comment by ZZYSonny** [comment]
OK I will try.

**Comment by ZZYSonny** [comment]
I've got text rendering (mostly) correct. The way I render text is

- For each embedded font, extract SVG path of unicode encodable glyph. Compose to SVG font and use `svg2ttf` to convert SVG font to TTF font.
- Use TextSpan whenever possible.
- For characters that cannot be encoded as unicodes, fallback to SVG path.

I am working on other elements of PDFs.

![image](https://user-images.githubusercontent.com/28442838/187065128-e9e3ab7d-78bc-4128-80ce-6e7f2f84d096.png)

**Comment by s3bk** [comment]
You can actually link most fonts directly in SVG.
CID fonts are a bit more difficult.

**Comment by ZZYSonny** [comment]
> You can actually link most fonts directly in SVG. CID fonts are a bit more difficult.

Not really. One can embed pfa fonts into a pdf, which is not supported by a modern browser. pfa (or any other) font -> svg font -> ttf font seems to be a shortcut, when there is no good tool to convert a pfa font on the web.

I am working on other drawing, and I got some transparency issue. I will open another issue in pdf-render.
